### PR TITLE
Allow overwriting STS config in ChainableTemporaryCredentials

### PR DESCRIPTION
**Pull request by workeitel**
STS uses the global endpoint per default which does not work for
"opt-in" regions. The `ChainableTemporaryCredentials` wrapper for
fetching credentials via `AssumeRole` or `GetSessionToken` does not
allow overriding the endpoint or set a different region.

This change will introduce another optional constructor parameter for
specifying the STS client configuration and forwarding it to the client
constructor.

In addition I remove the incorrect typing for the second
`masterCredentials` constructor parameter. The parameter needs to be
passed as a nested value in `options`.

issue: https://github.com/aws/aws-sdk-js/issues/2673

##### Checklist

- [X] `npm run test` passes
- [X] `.d.ts` file is updated
- [X] changelog is added, `npm run add-change`

### CODE REVIEW

File: lib/credentials/chainable_temporary_credentials.d.ts [hunk 1]
```diff
@@ -3,7 +3,7 @@ import {AWSError} from '../error';
 import STS = require('../../clients/sts');
 
 export class ChainableTemporaryCredentials extends Credentials {
-    constructor(options: ChainableTemporaryCredentials.ChainableTemporaryCredentialsOptions, masterCredentials?: Credentials);
+    constructor(options: ChainableTemporaryCredentials.ChainableTemporaryCredentialsOptions);
```
Comment (AllanFly120) [inline lib/credentials/chainable_temporary_credentials.d.ts]: Nice catch!

**Comment by AllanFly120** [comment]
To update the STS endpoint, you can actually do it as follows:
```javascript
const stsClient = chainableCredentials.service;
stsClient.setupEventListeners = (request) => {
    request.on('afterBuild', (req) => {
        request.httpRequest.endpoint = //endpoint you want
    })
}
```

**Comment by workeitel** [comment]
Hey @AllanFly120 . Thanks for the quick reply. I tried multiple approaches but could not get it working:

```
const credentials = new AWS.ChainableTemporaryCredentials({
  params: {
    DurationSeconds: 900
  }
});
```

# approach 1

As I originally suggested in the ticket https://github.com/aws/aws-sdk-js/issues/2673

```
credentials.service.config.endpoint = "sts.us-west-1.amazonaws.com";
```

did not set the endpoint
# approach 2

as you suggested

```
// @ts-ignore
credentials.service.setupEventListeners = request => {
  // @ts-ignore
  request.on("afterBuild", req => {
    request.httpRequest.endpoint = "sts.us-west-1.amazonaws.com";
  });
};
```

did not set the endpoint
# approach 3

as found somewhere else

```
// @ts-ignore
credentials.service.setEndpoint("sts.us-west-1.amazonaws.com");
```

did successfully override the endpoint - but not the region. So I end up with a `Credential should be scoped to a valid region, not 'us-east-1'.` exception.

I guess I'm doing something wrong.

**Comment by AllanFly120** [comment]
Hey @workeitel 

Sorry, I didn't make it clear in my previous response. The `endpoint` in SDK is an object. Here's its interface: https://github.com/aws/aws-sdk-js/blob/master/lib/endpoint.d.ts
So you should set the endpoint like:
```javascript
const stsClient = chainableCredentials.service;
stsClient.setupEventListeners = (request) => {
    request.on('afterBuild', (req) => {
        request.httpRequest.endpoint.host = "sts.us-west-1.amazonaws.com";
    })
}
```

**Comment by workeitel** [comment]
Thanks @AllanFly120. I tried your version but I assume you mean `setupRequestListeners` instead:

```
credentials.service.setupRequestListeners = request => {
  request.on("afterBuild", () => {
    request.httpRequest.endpoint = new AWS.Endpoint(
      "sts.us-west-1.amazonaws.com"
    );
  });
};
```

But it fails with:

```
"Host: sts.amazonaws.com. is not in the cert's altnames: DNS:sts.us-west-1.amazonaws.com, DNS:*.sts.us-west-1.vpce.amazonaws.com"
```

since the `Host` header is [already set](https://github.com/aws/aws-sdk-js/blob/master/lib/event_listeners.js#L182-L184). Adding

```
request.httpRequest.headers.Host = "sts.us-west-1.amazonaws.com"
```

helps (alternative use `build` state instead of `afterBuild`) but then it fails with

```
"Credential should be scoped to a valid region, not 'us-east-1'. "
```

I think for signing the request the AWS SigV4 requires the region as well since the region is part of the signature. If we override the endpoint with another region we need to update the region for the signature as well. https://docs.aws.amazon.com/general/latest/gr/sigv4_elements.html#sigv4_elements_auth_params

Overriding the region as well works but is not supported in typings: `request.httpRequest.region = "us-west-1";`

Final version:

```
import AWS = require("aws-sdk");

const credentials = new AWS.ChainableTemporaryCredentials({
  params: {
    DurationSeconds: 900,
    RoleArn:
      "arn:aws:iam::62..:role/..."
  }
});

function patchEndpoint(
  creds: AWS.ChainableTemporaryCredentials,
  endpoint: string,
  region: string
) {
  creds.service.setupRequestListeners = request => {
    request.on("build", () => {
      request.httpRequest.endpoint = new AWS.Endpoint(endpoint);

      // @ts-ignore
      request.httpRequest.region = region;
    });
  };
}

patchEndpoint(credentials, "sts.us-west-1.amazonaws.com", "us-west-1");

const client = new AWS.STS({
  region: "us-west-2",
  endpoint: "sts.us-west-2.amazonaws.com",
  credentials
});
client
  .getCallerIdentity()
  .promise()
  .then(console.log, console.log);
```

But I don't think thats a nice solution. At least the `region` should be in the [`httpRequest`](https://github.com/aws/aws-sdk-js/blob/master/lib/http_request.d.ts) typings. Or maybe making `.service` not [readonly](https://github.com/aws/aws-sdk-js/blob/master/lib/credentials/chainable_temporary_credentials.d.ts#L19) to allow overriding the STS client?

**Comment by AllanFly120** [comment]
@workeitel I see your point. I'm approving this PR.

**Pull request merged**